ZOOKEEPER-4956: Provide a HostProvider that uses DNS SRV record for dynamic server discovery

li4wang opened the pull request.

anmolnar reviewed (changes requested):
Where are the changes of `StaticHostProvider`?

anmolnar commented on zookeeper-server/src/main/java/org/apache/zookeeper/client/HostConnectionManager.java:
```diff
@@ -0,0 +1,391 @@
+ * HostConnectionManager handles the complex connection management and reconfiguration logic
+ */
+@InterfaceAudience.Private
+public final class HostConnectionManager {
```
This could be the common abstract class for both implementations. Doing it with inheritance might be simpler than this composition where you just forward the calls to Connection Manager.

li4wang replied:
I initially implemented it using inheritance but refactored to using the composition pattern for the following reasons:

Separation of Concerns: Cleanly separates connection management from DNS lookup logic, making each component focused on a single responsibility.

Extensibility: Makes it easy to add new provider types (Consul, etcd, etc.) without modifying existing code.

Testability: Makes it easy to test each component independently.

In my opinion, the minimal forwarding/delegation code is easily justified by for the improved flexibility, testability, and maintainability this approach provides. WDYT?

anmolnar replied:
So, basically what happened is that all of the "host providing" logic has been moved to `HostConnectionManager`, while the `HostProvideder` has become just the class whose sole responsibility is to be the source of server addresses.

In case of `StaticHostProvider`, this is a static list supplied in the constructor during instantiation, while for `DnsSrvHostProvider` it comes from DNS and gets periodically updated.

That seems to me fine.

li4wang commented:
> Where are the changes of `StaticHostProvider`?

I thought it would be better to check in the updated `StaticHostProvider` later in a separate PR, as the changes are code refactoring, so I took it out from this PR. 

However, I noticed that  `StaticHostProvider` is mentioned in the summary of this PR, let me just added it back to avoid any confusion.

li4wang commented on zookeeper-server/src/main/java/org/apache/zookeeper/client/DnsSrvHostProvider.java:
```diff
@@ -0,0 +1,383 @@
+     *
+     * @return the initialized Timer or null if refresh interval is 0 or negative
+     */
+    private Timer initializeDnsRefreshTimer() {
```
Good point on DNS caching. However, periodic refresh provides a critical operational benefit: immediate load balancing during scale-up.

Example: Scale 3→5 servers with 1000 clients:

Query on exhaustion: New servers sit idle, original 3 stay overloaded
Periodic refresh: Load redistributes within 60 seconds, immediate capacity benefit

Overhead is minimal: 1 DNS query/minute (typically <1ms from cache), single background thread.

For dynamic environments (Kubernetes, auto-scaling), this ensures scaling operations provide immediate capacity benefits through load balancing with minimal overhead.

Thoughts?

anmolnar commented on zookeeper-server/pom.xml:
```diff
@@ -190,6 +190,11 @@
+    <dependency>
+      <groupId>dnsjava</groupId>
+      <artifactId>dnsjava</artifactId>
+      <version>3.5.1</version>
```
The version should be defined in the root-level `pom.xml` file like for other third party libraries.

anmolnar reviewed (changes requested):
Just a few more comments, otherwise looks good to me.

anmolnar commented:
Please see my comments on the other pr #2321 
I think we should merge these patches into a single one.

li4wang commented on zookeeper-server/src/main/java/org/apache/zookeeper/client/HostProvider.java:
```diff
@@ -72,4 +72,21 @@ public interface HostProvider {
+     *
+     * @return the ConnectionType supported by this provider
+     */
+    default ConnectionType getSupportedConnectionType() {
```
Removed, not needed any more.

li4wang commented:
@anmolnar thanks a lot for your feedback. I've merged  https://github.com/apache/zookeeper/pull/2321 into this one and addressed all the comments. Would you mind taking a look at it? Thanks.

anmolnar commented on zookeeper-server/src/main/java/org/apache/zookeeper/client/DnsSrvHostProvider.java:
```diff
@@ -0,0 +1,337 @@
+        try {
+            final SRVRecord[] srvRecords = dnsResolver.lookupSrvRecords(dnsSrvName);
+            if (srvRecords.length == 0) {
+                LOG.warn("No SRV records found for DNS name: {}", dnsSrvName);
```
LOG.error()

anmolnar commented on zookeeper-server/src/main/java/org/apache/zookeeper/client/DnsSrvHostProvider.java:
```diff
@@ -0,0 +1,337 @@
+                        addresses.add(address);
+                    }
+                } catch (final Exception e) {
+                    LOG.warn("Failed to create address from SRV record {}: {}", srvRecord, e.getMessage());
```
LOG.error()

li4wang replied:
Removed.

The catch block is redundant, as `createAddressFromSrvRecord` already has try-catch that handles exceptions.

anmolnar commented on zookeeper-server/src/main/java/org/apache/zookeeper/client/DnsSrvHostProvider.java:
```diff
@@ -0,0 +1,337 @@
+            for (final SRVRecord srvRecord : srvRecords) {
+                try {
+                    final InetSocketAddress address = createAddressFromSrvRecord(srvRecord);
+                    if (address != null) {
```
You might want to LOG.error() the `null` value case here. Under what circumstances could the return value be null?

anmolnar commented on zookeeper-server/src/main/java/org/apache/zookeeper/client/DnsSrvHostProvider.java:
```diff
@@ -0,0 +1,337 @@
+            final int port = srvRecord.getPort();
+
+            if (port <= 0 || port > 65535) {
+                LOG.warn("Invalid port {} in SRV record for target {}", port, target);
```
LOG.error()

li4wang replied:
Updated

anmolnar commented on zookeeper-server/src/test/java/org/apache/zookeeper/client/DnsSrvHostProviderTest.java:
```diff
@@ -0,0 +1,205 @@
+
+            assertNotNull(addr1);
+            assertNotNull(addr2);
+            assertNotNull(addr3);
```
I think you could do more than just doing a null check. You might want to verify against the return value.

li4wang replied:
updated

anmolnar commented on zookeeper-server/src/test/java/org/apache/zookeeper/client/DnsSrvHostProviderTest.java:
```diff
@@ -0,0 +1,205 @@
+            assertEquals(3, hostProvider.size());
+
+            // Wait to ensure no background refresh occurs
+            Thread.sleep(100);
```
You can even wait for a second.

li4wang commented on zookeeper-server/src/test/java/org/apache/zookeeper/client/DnsSrvHostProviderTest.java:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    @Test
+    public void testRefreshIntervalNegativeUsesDefault() throws Exception {
```
Updated

anmolnar reviewed (approved):
lgtm. 🚢

li4wang commented:
@anmolnar Thanks a lot for reviewing the PR and providing the feedback!
